src/http-gateways/trustless-gateway.md

```diff
@@ -66,12 +74,137 @@ Below response types MUST to be supported:
 - [application/vnd.ipld.car](https://www.iana.org/assignments/media-types/application/vnd.ipld.car) – disables IPLD/IPFS deserialization, requests a verifiable CAR stream to be returned
 - [application/vnd.ipfs.ipns-record](https://www.iana.org/assignments/media-types/application/vnd.ipfs.ipns-record) – requests a verifiable :cite[ipns-record] (multicodec `0x0300`).
 
+## Request Query Parameters
+
+### :dfn[dag-scope] (request query parameter)
+
+Optional, `dag-scope=(block|entity|all)` with default value `all`, only available for CAR requests.
+
+Describes the shape of the DAG fetched the terminus of the specified path whose blocks
+are included in the returned CAR file after the blocks required to traverse
+path segments.
+
+- `block` - Only the root block at the end of the path is returned after blocks
+  required to verify the specified path segments.
+
+- `entity` - For queries that traverse UnixFS data, `entity` roughly means return
+  blocks needed to verify the terminating element of the requested content path.
+  For UnixFS, all the blocks needed to read an entire UnixFS file, or enumerate a UnixFS directory.
+  For all queries that reference non-UnixFS data, `entity` is equivalent to `block`
+
+- `all` - Transmit the entire contiguous DAG that begins at the end of the path
+  query, after blocks required to verify path segments
+
+When present, returned `Etag` must include unique prefix based on the passed scope type.
+
+### :dfn[entity-bytes] (request query parameter)
+
+Optional, `entity-bytes=from:to` with the default value `0:*`, only available for CAR requests.
+Serves as a trustless form of an HTTP Range Request.
+
+When the terminating entity at the end of the specified content path can be
+interpreted as a continuous array of bytes (such as a UnixFS file), returns
+only the minimal set of blocks required to verify the specified byte range of
+said entity.
+
+Allowed values for `from` and `to` are positive integers where `to` >= `from`, which
```

Comment:
> where `to` >= `from`, which

Similar (same?) question as the one below about `from` and `to` resolving to a zero byte range. Because `=` is allowed here, what does that mean in the `=` case?

Separate question: what does it mean for them not to be "allowed"? We could reject obviously bogus queries with an error at the top end of the request (e.g. `entity-bytes=100:10` - error, bad range). But what about where you don't know the range falls within this "allowed" ruleset until you have the file details—this uncertainty is created by negative byte ranges which could mean `to` < `from` once you know how long the file is. Is this an error? Or do we massage values for the user?

Reply:
There is `=` because HTTP Ranges are inclusive (`0-0` means one byte, see [RFC9110#14.1.2](https://httpwg.org/specs/rfc9110.html#rfc.section.14.1.2)). The `entity-bytes` offsets follow the same notation to avoid translation and one-off errors.  I've added explicit reference to this RFC so its more clear how offsets work.

To create zero-range you would have to use negative values from the end, and match file length perfectly. Doable, but also something that requires knowledge of the DAG, thats why I think we should return `entity-dag

For `to < from` cases, 608e7b7 added:

> A Gateway SHOULD return an HTTP 400 Bad Request error when the requested range cannot be parsed as valid offset positions.

I think if you know upfront the range is bogus, and the client has all information to come up to the same conclusion themselves (to < from), it is ok to return error.  

But in more nuanced cases where client may not have enough knowledge of the DAG,  we need to return  at least the root block.

Reply:
OK, so what I'm working on right now gets to the question of bad ranges.

* Selector spec update: https://github.com/ipld/ipld/pull/289
* Selector implementation for negatives: https://github.com/ipld/go-ipld-prime/pull/530

What I'm going with is this: if we get to the range match, and then perform whatever calculations need to be made to adjust `To` and `From` and we find that it specifies a range entirely outside of the bytes we have available, then it's not going to match anything. This means that you'll get the root block and nothing else. But you also won't get an error condition (no dirty stream termination)—although you could figure that out yourself with the data you get back.

Some edge conditions:

* `To` beyond the end of the file is fine (and I need to use this fact to convert `[From,To]` in this spec to `[From,To)` in the impl.)
* `From` before the start of the file is not fine and won't match anything (although perhaps it could be? Maybe it doesn't matter if you end up matching bytes anyway. I might change this but input welcome!)
* `To`<`From` is not fine and won't match anything. If you provide positive values and `To`<`From` at the beginning then we can `400 Bad Request` on that. But we can't do that if you provide negatives, just no blocks beyond the root.

Reply:
Update: I'm going to be more permissive with the `From` value underflowing. So the rules end up being like this:

* Incoming request `From>0 && To>0 && From>To` - `400 Bad Request`
* After resolving negative `From` and/or `To` values once we have the file data:
  - Where `From<0`, `From` is adjusted to `0`
  - Where `To>=Length`, `To` is adjusted to `Length-1`
  - Where `From>To`, nothing is returned.
  - Where `To<0`, nothing is returned.
  - Where `From>=Length`, nothing is returned.
  - Where `From==To`, the block at that byte is returned.
  - Normal rules apply for other ranges.

(Although these rules are mediated through a `[From,To]` -> `[From,To)` translation layer which is slightly awkward)

Reply:
@rvagg it's worth clarifying to those reading the above that the rule set described is applied AFTER any negative indexes are converted to their equivalent index from the start of the bytes.

Reply:
@rvagg @hannahhoward  by "nothing is returned"  you mean truly empty car (useless to the client -- unable to tell what went wrong)  or equivalent of `dag-scope=block` (path+root, allowing client to understand what went wrong with their request)?

The spec states the latter – any concerns around that?

> In more nuanced error scenarios, a Gateway MUST return a valid CAR response that includes enough blocks for the client to understand why the requested  `entity-bytes` was incorrect or why only a part of the requested byte range was
returned:
> [..]
> If the requested `entity-bytes` resolves to a zero-length range or falls   fully outside of the entity's bytes, the response is equivalent to   `dag-scope=block`.

Reply:
by "nothing is returned" I mean none of the child blocks for a sharded file - you get everything up to the parent block of the sharded file and nothing beyond that; so still a valid CAR, and technically you could validate for yourself why it failed by rerunning the byte range calculation locally, so this seems fine I think?